### ISSUE TITLE
Fetching unneccsary columns when projecting to a new class

### ISSUE THREAD

**Issue by AradAral** [ChilliCream/hotchocolate#2373]
Following https://github.com/ChilliCream/hotchocolate/issues/2365#issuecomment-697727508
I have `Book` and `Author` domain classes + `BookDto` and `AuthorDto` classes, I want to expose an `IQueryable<BookDto>` for the `books` field on my `query` root type.
So here is the `GetBooks` method on the `Query` class: It simply projects `Book` into `BookDto` and returns an `IQueryable<BookDto>` so that Hot Chocolate could then do its magic on it.
```csharp
[UseSelection]
public IQueryable<BookDto> GetBooks([Service]AppDbContext dbContext)
{
    return dbContext.Books.Select(book => new BookDto
    {
        Name = book.Name,
        ...
        Author = new AuthorDto
        {
            Id = book.Author.Id,
            Name = book.Author.Name,
        }
    });
}
```
Now, when you run the following query:
```graphql
{
  books {
    name
    author {
      name
    }
  }
}
```
You would probably expect the following SQL to ultimately get generated:
```sql
SELECT [b].[Name], [a].[Name]
      FROM [Books] AS [b]
      INNER JOIN [Authors] AS [a] ON [b].[AuthorId] = [a].[Id]
```
However, instead, Hot Chocolate causes this one to get generated by EF Core:
```sql
SELECT [b].[Name], [a].[Id], [a].[Name], [a].[Age]
      FROM [Books] AS [b]
      INNER JOIN [Authors] AS [a] ON [b].[AuthorId] = [a].[Id]
```
Which retrieves all the fields on the author (including `Age` and `Id` in this case), even though the query requested only the `Name`.

At first, I thought this must be an EF Core issue, but then I executed the following code:
```csharp
var queryableDto = dbContext.Books.Select(book => new BookDto
{
    Name = book.Name,
    Author = new AuthorDto
    {
        Id = book.Author.Id,
        Name = book.Author.Name,
        Age = book.Author.Age
    }
});
// The code that you would expect Hot Chocolate will generate for the above GraphQL query:
queryableDto.Select(bDto => new
{
    Name = bDto.Name,
    Author = new
    {
        Name = bDto.Author.Name,
    }
}).ToList();
```
And I realized it works as expected, and generates the following SQL which only selects the requested column (author name in this case):
```sql
SELECT [b].[Name], [a].[Name]
      FROM [Books] AS [b]
      INNER JOIN [Authors] AS [a] ON [b].[AuthorId] = [a].[Id]
```

Therefore, this must be related to Hot Chocolate, I believe.
This is really a serious problem for situations like mine when we don't want to expose `IQueryable<TheEntityFrameworkModelClass>` but rather a different type.

I'm using Hot Chocolate version 10.5.2
And I've tried this with both EF Core version 3.1.8 and 5.0.0 RC.

**Comment by PascalSenn** [comment]
i will track this down in the new implementation for 11. 

It should just work, because we prjoect like this: 

```graphql
{
  books {
    name
    author {
      name
    }
  }
}
```

```csharp
.Select(x=> new BookDto
            {
                Name = x.Name,
                Author = new AuthorDto
                { 
                    Name = x.Author.Name,
                }
            })
```

**Comment by AradAral** [comment]
@PascalSenn Ok, great to hear. Then we'll have to wait for v11 before we can deploy our app so that this issue is fixed until then. Thanks.
> It should just work, because we prjoect like this:

Well, as I mentioned [here](https://github.com/ChilliCream/hotchocolate/issues/2365#issuecomment-697727508), running the code that you said Hot Chocolate would generate from the GraphQL query:
```
queryableDto.Select(bDto => new
{
    Name = bDto.Name,
    Author = new
    {
        Name = bDto.Author.Name,
    }
}).ToList();
```
would NOT generate the SQL that Hot Chocolate causes to be generated. I talked about this in the [original comment](https://github.com/ChilliCream/hotchocolate/issues/2373#issue-707805666) and also [here](https://github.com/ChilliCream/hotchocolate/issues/2365#issuecomment-697727508). So I think that's probably not exactly how the projection works in Hot Chocolate.
And I feel like this might be a little mistake or something somewhere, it's probably not that big of a deal on your side, I don't know I'm just guessing, it is, however, a very big deal when it comes to our side! Since all the data of a related entity gets fetched unnecessarily. So, I'd appreciate it if you look into this in the coming days! Thank you so much.

**Comment by PascalSenn** [comment]
Does this query generate the desired result? 
```
queryableDto.Select(x=> new BookDto
            {
                Name = x.Name,
                Author = new AuthorDto
                { 
                    Name = x.Author.Name,
                }
            }).ToList();
```

**Comment by AradAral** [comment]
@PascalSenn Yes, that query would generate the following SQL:
```
SELECT [b].[Name], [a].[Name]
      FROM [Books] AS [b]
      INNER JOIN [Authors] AS [a] ON [b].[AuthorId] = [a].[Id]
```
which is perfect. It only retrieves the name of the book and the name of the author, as requested.

However, sending this GraphQL query to our Hot Chcolate server:
```
{
  books {
    name
    author {
      name
    }
  }
}
```
would generate this SQL instead:
```
SELECT [b].[Name], [a].[Id], [a].[Name], [a].[Age]
      FROM [Books] AS [b]
      INNER JOIN [Authors] AS [a] ON [b].[AuthorId] = [a].[Id]
```
which as you can see is unnecessarily fetching the `Id` and `Age` of the author, though they were not queried.

**Comment by PascalSenn** [comment]
Yes a repro would be great

Are there any custom resolvers on author?

**Comment by AradAral** [comment]
@PascalSenn Sure! I'll do it and let you know.
And no, nothing, it's a just class with a handful of properties. I've added nothing HotChocolate-specific to it at all.

**Comment by AradAral** [comment]
@PascalSenn Here you go, sir: https://github.com/AradAral/AspNetCoreGraphQL

**Comment by AradAral** [comment]
@PascalSenn:

![Untitled Project](https://user-images.githubusercontent.com/26527405/94172100-ef390200-fe9e-11ea-8cc0-0ab2932fe97f.gif)

**Comment by PascalSenn** [comment]
@AradAral  cool, the repro is definitely helpful. 
Checked the configuration didn't see anything obvious wrong with it.

**Comment by AradAral** [comment]
@PascalSenn Oh! I just noticed something else which I think might be important:
I removed the `UseSelection` middleware from the `GetBooks` method, but it still gave me the author data if I queried it, which normally wouldn't happen unless you use the `UseSelection` middleware.
Do you think that can tell us anything about the issue?

**Comment by PascalSenn** [comment]
Did you also remove it from the resolver declared in the QueryType?

It is probably always requested in this case, because Select statement that maps from object to dto projects it already

**Comment by AradAral** [comment]
Yes, I actually removed the whole `QueryType` class and it's just now the `Query` class, without the UseSelection attribute on top of the resolver.
Now that the `UseSelection` middleware is not present, I just noticed that it retrieves the author data even I don't query it, but that wasn't the case when the `UseSelection` middlware **was** present, in that case it would only retrieve the author data (but the whole author data) if I requested at least one field from it.

**Comment by AradAral** [comment]
@PascalSenn isn't there any way to inspect the expression that Hot Chocolate passes to the `Select` method?

**Comment by AradAral** [comment]
@PascalSenn I FOUND OUT!
I think I just found out what's causing the problem:
By building an expression visitor and inspecting the expression that Hot Chocolate passes to the `Select` method, I just realized that it (the expression Hot Chocolate generates) looks like this:
```
e => new BookDto()
    {
        Name = e.Name,
        Author = e.Author == null ? default(AuthorDto) : new AuthorDto() { Name = bDto.Author.Name }
    }
```
The important part is where it's first checking whether `Author` is null or not, that is exactly what's causing the whole author object to be loaded from the database. This works without any problem with EF Core entity classes, but not with DTOs like this.

Is there any to do something about this? Can't I disable null checks or something?

**Comment by AradAral** [comment]
Based on the analysis, I'll close this issue.
For anyone who's curious, I eventually ended up writing a custom expression visitor that replaces any occurrence of `e.Author == null` that Hot Chocolate generates to `e.Author.Id == null`, so as to avoid the aforementioned over-etching problem.

**Comment by PascalSenn** [comment]
@AradAral  sorry for not mentioning this, this should be fixed in the newest preview

**Comment by AradAral** [comment]
@PascalSenn Hi!
I didn't know that, what change did you make to fix this? I thought there was nothing you could do about it.

**Comment by PascalSenn** [comment]
Sorry i messed something up. 
This could still be an issue

**Comment by PascalSenn** [comment]
we just fixed UseProjections to do this:
```        Author = e.Author == null ? default(AuthorDto) : new AuthorDto() { Name = e.Author.Name }```

I need to check what it does with reprojection

**Comment by AradAral** [comment]
@PascalSenn 
But isn't `Author = e.Author == null ? default(AuthorDto)...` what HC already did in the previous versions? What has changed exactly?

**Comment by PascalSenn** [comment]
Yes, but UseProjection did something different

**Comment by AradAral** [comment]
@PascalSenn Aha! Okay, got it!
No problem ;)

**Comment by gojanpaolo** [comment]
It looks like there's still an issue on deeply nested entities.

```c#
[UseSelection]
public IQueryable<BookDto> GetBooks([Service]AppDbContext dbContext)
{
    return dbContext.Books.Select(book => new BookDto
    {
        Name = book.Name,
        ...
        Author = new AuthorDto
        {
            Id = book.Author.Id,
            Name = book.Author.Name,
            Address = new AddressDto
            {
                Street = book.Author.Address.Street
            }
        }
    });
}
```

where the `Address` is still joined when we only query like so

```graphql
{
  books {
    name
    author {
      name
    }
  }
}
```

**Comment by AradAral** [comment]
@gojanpaolo Hi!
It's actually not just on deeply nested objects. Whenever you have anything like `[Something] = new [Something]Dto { }` in your projection that you return in your resolvers, even one field getting requested by the GraphQL client on that `[Something]Dto` causes the whole object to be retrieved from the database. I explained why that happens [here](https://github.com/ChilliCream/hotchocolate/issues/2373#issuecomment-699232133).

The reason why in your case `Address` is being retrieved even though you're only querying `Author.Name` is because `Author` is because, again, the whole `Author` object is getting retrieved, which means everything inside of it is getting retrieved, including every nested object, like `Address` in your case.

**Comment by gojanpaolo** [comment]
I was expecting the projection to behave similar to
```c#
var dto = 
   dbContext.Books.Select(book => new BookDto
    {
        Name = book.Name,
        ...
        Author = new AuthorDto
        {
            Id = book.Author.Id,
            Name = book.Author.Name,
            Address = new AddressDto
            {
                Street = book.Author.Address.Street
            }
        }
    });

var result =
    dto.Select(bookDto => new
    {
        Name = bookDto.Name,
        Author = new
        {
            Name = bookDto.Author.Name
        }
    })
    .ToList();
```

which EF correctly translates to the SQL query I expected, i.e. no join on the `Address`.

I believe this is something we can fix either in hotchocolate/`UseProjection` itself, or in my graphql layer code.

**Comment by jannikbuschke** [comment]
Im also experiencing overfetching:

the query
```graphql
users { id }
```
works fine and produces the following SQL
```sql
SELECT [u].[Id]
FROM [User] AS [u]
```
whereas this query
```graphql
users {
  id
  author { name }
}
```
unecessary loads more data:
```sql
SELECT [u].[Id], [a].[Id], [a].[Name], [a].[UserId], [b].[Id], [b].[AuthorId], [b].[Title]
FROM [User] AS [u]
LEFT JOIN [Author] AS [a] ON [u].[Id] = [a].[UserId]
LEFT JOIN [Book] AS [b] ON [a].[Id] = [b].[AuthorId]
ORDER BY [u].[Id], [a].[Id], [b].[Id]
```

```csharp
[UseProjection]
public IQueryable<User> Users([Service] DataContext ctx)
{
     return ctx.User
         .Include(v => v.Author)
         .ThenInclude(v => v.Books);
}
```

Using efcore 5 and hot chocolate 11

**Comment by brickscrap** [comment]
I see this issue was closed alongside a load of AutoMapper tickets, but this doesn't seem related to AutoMapper? I'm having the same issue here using projections, where all fields are requested in the SQL query, despite only requesting one field of a sub-object

**Comment by PascalSenn** [comment]
@brickscrap  can you share your resolver and query?

**Comment by brickscrap** [comment]
Ignore me - despite trying to figure it out for hours before commenting here, 10 minutes after posting I found the problem. I had UseProjection enabled on the parent object, and this seemed to be causing issues. Removed it, and it's all working as expected now.

Sorry for bothering you

**Comment by nollidnosnhoj** [comment]
I am getting the same issue where it is fetching all the columns in child instead of what is selected.

query

```csharp
public class Query
{
    [UseProjection]
    public IQueryable<PersonDto> GetPersons([Service] SomeDbContext dbContext)
    {
        return dbContext.Persons.Select(person => new PersonDto
        {
            Id = person.Id,
            FirstName = person.FirstName,
            LastName = person.LastName,
            Guardian =new GuardianDto
            {
                Id = person.Guardian.Id,
                FirstName = person.Guardian.FirstName,
                LastName = person.Guardian.LastName
            }
        });
    }
}
```

dtos (objects)

```csharp
public class PersonDto
{
    public int Id { get; set; }
    public string FirstName { get; set; } = null!;
    public string LastName { get; set; } = null!;
    public GuardianDto? Guardian { get; set; } = null!;
}

public class GuardianDto
{
    public int Id { get; set; }
    public string FirstName { get; set; } = null!;
    public string LastName { get; set; } = null!;
}
```

selections

```graphql
query {
  persons {
    id
    firstName
    guardian {
      id
      firstName
    }
  }
}
```

sql

```sql
SELECT "p"."Id", "p"."FirstName", "g"."Id", "g"."FirstName", "g"."LastName"
      FROM "Persons" AS "p"
      LEFT JOIN "Guardians" AS "g" ON "p"."GuardianId" = "g"."Id"
```

**Comment by AradAral** [comment]
For anyone still having this problem, I explained the reason why it's happening in [this comment](https://github.com/ChilliCream/hotchocolate/issues/2373#issuecomment-699232133).

It would be cool if Hot Chocolate provided a configuration setting or something on its projection engine to allow the developer to choose how the existence or non-existence of a single related entity would be determined in the projection.

**Comment by AradAral** [comment]
So let me reopen this issue.

**Comment by PascalSenn** [comment]
The reason why this check is there, is to ensure correct nullability in the projection. If this check is not there the projection will fail if you run the query in memory. 
What format would work for you? just enforcing the format?

>I think since this pattern of having "intermediary projections" (like DTOs) is rather common

It is actually not. Most of the time you do not need these intermediary objects. Normally the Types are your DTO. If you do 1:1 mappings you essentially write a DTO for nothing. The type control the visibility of fields and such. A clean Domain layer leads in most cases to a clean GraphQL schema.

**Comment by AradAral** [comment]
> The reason why this check is there, is to ensure correct nullability in the projection. If this check is not there the projection will fail if you run the query in memory.

I understand the reason for that check. I'm not saying it should be removed.

> It is actually not. Most of the time you do not need these intermediary objects. Normally the Types are your DTO.

I beg to differ, Pascal. The cleanness of the underlying entity classes has nothing to do with whether or not you would need intermediary objects.
Consider the following entity structure:
```csharp
public class Product
{
    public int Id { get; set; }
    public string Title { get; set; }
    public ProductRating Ratings { get; set; }
}

public class ProductRating
{
    public int ProductId { get; set; }
    public Product Product { get; set; }
    public byte Rating { get; set; }
}
```
In this case, you might want the type `Product` in your GraphQL schema to have a field called `averageRating`. You don't have such a field on your entity classes and you shouldn't, as this is "computed" field.

This is where intermediary projections would be necessary:
```csharp
dbContext.Products.Select(p => new ProductDto
{
    AverageRating = p.Ratings.Average(r => r.Rating),
});
```

Apart from computed fields, there's also a lot of low-level details in your entity structure that would most likely want to hide from the user of your GraphQL API. An example of that would be foreign key properties like `ProductId`, etc.

Back to the overfetching issue as a result of the null check in the projection: What I've personally done is I've written an expression visitor that replaces every occurrence of those null checks, with null a check **on the ID of the object**.
So it turns the following (which is what HC generates):
```csharp
Author = e.Author == null ? default(AuthorDto) : new AuthorDto() { Name = e.Author.Name }
```
into:
```csharp
Author = e.Author.Id == null ? default(AuthorDto) : new AuthorDto() { Name = e.Author.Name }
```
With this, the overfetching issue is sovled and EF Core generates exactly the right SQL.

**Comment by tobias-tengler** [comment]
Everything you mentioned about computed fields and ignoring sensitive members is already solved by Hot Chocolate's type system. You can easily add new computed fields using Type Extensions and hide members from your entities using various APIs in Hot Chocolate. This does not require intermediary objects that need to be involved in the projection.

**Comment by PascalSenn** [comment]
> In this case, you might want the type Product in your GraphQL schema to have a field called averageRating. You don't have such a field on your entity classes and you shouldn't, as this is a "computed" field 
> An example of that would be foreign key properties like ProductId, etc. You don't want these to appear in your GraphQL schema.

Exactly, computes fields and ignores. This is what types can do too. 
```csharp
// Annotation Based example, but same thing works also in code or schema first.
[ExtendObjectType(typeof(Product), , IgnoreFields = new []{nameof(Product.ProductId)})]
public class ProductExtensions {
    public double GetAveragePrice(....) =>... 
}
```
I left the service injection up to your imagination (...). You can either have a domain method that calculates the rating, you can data load the ratings, you can do both of the above and many other things. 

Don't get me wrong, I think that there are some cases where a projection could be beneficial. Though what I see most often with DTO's is that it ends up in mostly 1:1 mappings and often a resolver could do the job when there is some field that you need. 

> Author = e.Author.Id == null ? default(AuthorDto) : new AuthorDto() { ... }

The problem here is that we do not know the key. The key could also be `e.Author.SomeForeignKeyId`. 
@michaelstaib  Thoughts on new syntax to identify the key?

**Comment by AradAral** [comment]
> Everything you mentioned about computed fields and ignoring sensitive members is already solved by Hot Chocolate's type system. You can easily add new computed fields using Type Extensions and hide members from your entities using various APIs in Hot Chocolate. This does not require intermediary objects that need to be involved in the projection.

@tobias-tengler Regarding ignores, yes. But computed fields, not really. Because with type extensions, the new fields that you write are not added to the projection expression, which means they would essentially send additional SQL roundtrips, a significant disadvantage, I would say, especially if you have a relatively large number of computed fields. With DTOs, you don't have this problem.

There's also the issue of developer ergonomics, I would argue it's cleaner to have classes that directly correspond to your resulting GraphQL types. Again, with DTOs, you'll get this benefit too. But with type extensions and ignores, this becomes a bit more complex to look at and harder to immediately parse in your head. But this can be subjective.

Tell me if I'm missing something.

> The problem here is that we do not know the key. The key could also be e.Author.SomeForeignKeyId

I think a simple method on `IObjectFieldDescriptor` would do:
```
descriptor.Field(o => o.Id)
    .Key();
```
Whenever the key is specified like this, Hot Chocolate projection engine would do the null checks on this field.
What do you think?

**Comment by PascalSenn** [comment]
```csharp
descriptor.Field(o => o.Id)
    .Key();
```

**Comment by PascalSenn** [comment]
```csharp
descriptor.Field(o => o.Id).Key();
```
This would be the easiest solution indeed, but it's bad in the bigger pictues.

1. You have to add the key for projections to work properly
2. What happens if you select two keys
3. `Key()` is a pretty strong term, that could be used for many different use cases (stitching, execution engine etc.)

When we identity keys on types, we can only do this in respect to the bigger picture. There are a lot of usecases where we need keys, so this API is not a easy decision.

**Comment by AradAral** [comment]
> You have to add the key for projections to work properly

Well, yes, of course this is necessary only when you have intermediary projections. I think a note in the docs in the [projections section](https://chillicream.com/docs/hotchocolate/fetching-data/projections) pointing this out would suffice.

> What happens if you select two keys

Maybe an exception? Unless you want to support composite keys, in which case I'm not sure.

> When we identity keys on types, we can only do this in respect to the bigger picture. There are a lot of usecases where we need keys, so this API is not a easy decision.

Sure, currently we don't have this key thing as far as I know, so I think we could start with this particular use case and then also gradually use it for other things (like stitching and stuff, as you mentioned) whenever the need came up?

But I'm not in a hurry since I have developed a workaround that's working for now, but it's a little clunky so I think it would make sense it HC supports this.

**Comment by alienwareone** [comment]
> ...which means they would result in additional SQL roundtrips, a significant disadvantage, obviously, especially if you have a relatively large number of computed fields.

I wanna share an other case I'm using a lot in my projects.

```
public class Product
{
    public int Id { get; set; }
    public int BrandId { get; set; }
    public string Name { get; set; }
    public Brand Brand { get; set; }
}

public class Brand
{
    public int Id { get; set; }
    public int ProductId { get; set; }
    public string Name { get; set; }
    public ICollection<Product> Products { get; set; }    
}
```

```
public class Query
{
    [UseProjection]
    public async Task<IQueryable<BrandDto>> GetBrands(
        [Service] ApplicationDbContext dbContext,
        [Service] IProductFilterService productFilterService,
        ProductFiltersInputModel productFilters)
    {
        var productFilterExpression = await productFilterService.BuildProductFilterExpressionAsync(productFilters with { BrandIds = null });
        return dbContext.Brands.Select(brand => new BrandDto
        {
            Id = brand.Id,
            Name = brand.Name,
            TotalProductCount = dbContext.Products.Count(product => product.BrandId == brand.Id)
            FilteredProductCount = dbContext.Products.Where(product => product.BrandId == brand.Id).Count(productFilterExpression),
        });
}
```
Of Course this approach isn't that bad but a way to configure this in my ObjectType would be awesome.
```
public class BrandType : ObjectType<Brand>
{
    protected override void Configure(IObjectTypeDescriptor<Brand> descriptor)
    {
        descriptor.Field("FilteredProductCount")
            .Type<IntType>()
            .Argument("productFilters", x => x.Type<ProductFiltersInputModel>())
            .Computed(async context =>
            {
                var productFilters = x.ArgumentValue<ProductFiltersInputModel>("productFilters");
                var productFilterService = context.Service<IProductFilterService>();
                var dbContext = context.Service<ApplicationDbContext>();
                var productFilterExpression = await productFilterService.BuildProductFilterExpressionAsync(productFilters with { BrandIds = null });
                LambdaExpression selector = (Brand brand) => dbContext.Products.Where(product => product.BrandId == brand.Id).Count(productFilterExpression);
                return selector;
            });
    }
}
```
With this approach we could fetch this computed fields from other ObjectType's.

**Comment by kresimirlesic** [comment]
@PascalSenn first of, thank you and the team for creating an awesome library, I am really enjoying it so far.

Secondly, I did a bit of an investigation of the original bug report and I think I have nailed down the issue.

The issue begins in [this line of code](https://github.com/ChilliCream/hotchocolate/blob/447bf4fb50c1ca55562a90fef31fc15c7411d1e5/src/HotChocolate/Data/src/Data/Projections/Expressions/Handlers/QueryableProjectionFieldHandler.cs#L99) as others have already pointed out.

However, removal of "NotNullAndAlso" is not enough to resolve this issue, unfortunately the issue is in [this line of code as well](https://github.com/ChilliCream/hotchocolate/blob/447bf4fb50c1ca55562a90fef31fc15c7411d1e5/src/HotChocolate/Data/src/Data/Projections/Expressions/Handlers/QueryableProjectionFieldHandler.cs#L89) because it is just taking the whole property, without considering what parts of the property are selected.

Basically, in the example of the original bug report what hotchocolate is doing is this:
`Author = x.Author`

While it should actually be doing something like this (hopefully?):
`Author = new AuthorDto { Name = x.Author.Name }`

Here is a screenshot of the somewhat expected code change to support this:
![image](https://user-images.githubusercontent.com/34268109/162713159-20ea5954-78ed-456f-b469-1be497c300ae.png)

--

Obviously this is a naive solution (it does not take into consideration nested properties?), but this does resolve the issue of overselecting the columns from the database. 

Do you think some way of "opting into" this feature would be the way to go (at least for level 1 properties)? 
Is this something that you and the team are planning to fix at all? 

Please note that this issue is quite larger than it actually seems at first.

Thankfully in the example above, only the "Name" column was overselected, however it is highly possible for more complex applications to have something like `SomeProp = x.Nav1.Nav2.Nav3.Nav4.SomeColumn`  which would result in 4 extra joins, even when you are not asking for the `SomeProp` in your query.  Not to mention the subselects, computed stuff, TVF and all those things that could be binding to the `SomeProp`. I believe you can see how undesirable this is.

I am looking forward to the discussion and getting this thing resolved!

**Comment by PascalSenn** [comment]
@kresimirlesic  a few lines above you see the memberinit variable, this is the variable that does the projection for the properties 
'x.Author = new Author { Name = x.Author.Name }'

**Comment by kresimirlesic** [comment]
@PascalSenn My apologies, you are completely right.
I have missed that part while I was trying to figure out what is going on :) 

So to continue on the discussion.... 

Do you think opting out of the NotNullAndAlso would be a good and simple alternative to the Key() API which was discussed above or do you already have a plan on how to resolve this in the near future?

**Comment by kresimirlesic** [comment]
@PascalSenn are there any plans to resolve this issue? 

Once you remove the NotNullAndAlso everything works as expected with regards to EF. 

It would be really great if this was supported out of the box as this is really the behavior that most of the people would expect.

**Comment by PascalSenn** [comment]
So removing `NotNullAndAlso` does not fix the issue:
![image](https://user-images.githubusercontent.com/14233220/163629430-eddc882b-4958-496f-881e-e6beab2a3596.png)
![image](https://user-images.githubusercontent.com/14233220/163629480-321a1a48-7104-4532-8964-e1775773af9f.png)
```sql
SELECT "b"."Title", "a"."Id", "a"."Name"
FROM "Books" AS "b"
INNER JOIN "Authors" AS "a" ON "b"."AuthorId" = "a"."Id"
``` 
this is the pull request: https://github.com/ChilliCream/hotchocolate/pull/4970
We are open for suggestions

**Comment by AradAlvand** [comment]
@kresimirlesic 
> While it should actually be doing something like this (hopefully?):
Author = new AuthorDto { Name = x.Author.Name }

I'm not sure how that would solve this issue? I thought we established the fact that a null check is necessary.

The only way this problem could be reliably solved is by switching the null check from the entity itself to its key. As I explained in [this comment](https://github.com/ChilliCream/hotchocolate/issues/2373#issuecomment-1032544128).

Tell me if I'm missing something.

**Comment by AradAlvand** [comment]
@PascalSenn Do you still believe adding a `.Key()` method to `IObjectFieldDescriptor` — or something similar — isn't feasible?
Correct me if I'm wrong but I find it a neat solution.

**Comment by kresimirlesic** [comment]
@AradAlvand 
> I'm not sure how that would solve this issue? I thought we established the fact that a null check is necessary.

Null check is not necessary for EF, only if the query gets executed in the memory for whatever reason, then we would get the exception.

Since Pascal said that the .Key() API might be bad in the bigger picture, I just wanted to provide a quick and working alternative to this problem while the team figures our if they want to go with the .Key() API or something else.

**Comment by AradAlvand** [comment]
@kresimirlesic Okay, that makes sense. But apart from whether or not this workaround would work or not — cuz @PascalSenn said in this comment that it's apparently not working? — I still think it's better if we go with a 100% solution, as you described, which is the `Key()` API. Because removing the null check would be problematic even for EF in cases where the relationship is optional and not required. In our book/author example, imagine if we had some books that weren't linked to an author, without the null check we wouldn't know.

**Comment by AradAlvand** [comment]
@PascalSenn Don't you think we could add the `.Key()` think right now to solve this issue and then add further functionality on top of it in the future if need be, Pascal? Would there be anything wrong with that?

**Comment by PascalSenn** [comment]
@AradAlvand  The approach @kresimirlesic  suggested is indeed working. @kresimirlesic  found an issue in my pull request. Very well spotted @kresimirlesic!.

**Comment by PascalSenn** [comment]
@AradAlvand 
>@PascalSenn Don't you think we could add the .Key() thing right now to solve this issue and then add further functionality on top of it in the future if need be? Would there be anything wrong with that, Pascal?

Because it doesnt solve the issue? 
```csharp
Author = e.Author.Id == null ? default(AuthorDto) : new AuthorDto() { ... }
```
This will never be true ` e.Author.Id == null`. Unless the ID of author is null and i doubt that this is the intend

**Comment by AradAlvand** [comment]
> Because it doesnt solve the issue?

It does. As I said, that's what I'm currently doing and it's working.

**Comment by AradAlvand** [comment]
I can create a quick repro for you, if you want.

**Comment by PascalSenn** [comment]
@AradAlvand  I added a unittests with a small playground in the pr #4970. Can you add it there?

**Comment by aradalvand** [comment]
(comment for stalebot)

**Comment by kresimirlesic** [comment]
@PascalSenn @michaelstaib has there been any progress on this issue?

**Comment by dotarj** [comment]
@aradalvand In [your comment on 10 November 2022](https://github.com/ChilliCream/hotchocolate/issues/2373#issuecomment-724886145) you mentioned that you wrote a custom expression visitor replacing occurrences of `e.Author == null`. I have a related problem in which I also need to modify the expression tree. Could you share a snippet of how you wrote this expression visitor and where you registered it?

**Comment by aradalvand** [comment]
@dotarj https://gist.github.com/aradalvand/9b70e8bb455f5398affba610f2a25625

**Comment by kresimirlesic** [comment]
@michaelstaib any plans to get this one on the roadmap now that the v13 is live? :)

**Comment by aradalvand** [comment]
Hey, any updates on this? Won't you consider it for v14?

**Comment by michaelstaib** [comment]
> @aradalvand I added a unittests with a small playground in the pr https://github.com/ChilliCream/graphql-platform/pull/4970. Can you add it there?

I think this was the latest state ... 

From our side this has no traction at the moment.

From my perspective on the GraphQL core I do not want a `Key` descriptor. I also think its not necessary .... 

Can you follow up on pascals question?

**Comment by aradalvand** [comment]
@michaelstaib 
> I think this was the latest state ...

Not really, [this](https://github.com/ChilliCream/graphql-platform/issues/4995#issuecomment-1107538143) was the latest state:

> This is exactly why we are so opposed to the key solution. I will talk to Shay from the EF team about this. HotChocolate.Data will get a major update for 13 and we have these issues on our list to solve.

I'd be curious to know if you did actually speak to Shay about this and if so then what was the result?

Also, #2365, if implemented, could most likely solve this whole class of problems.

**Comment by michaelstaib** [comment]
We did at some point discuss projections. But the focus was more on Keyset pagination and nested queries. Overall we did not revamp projections for Hot Chocolate 13 and postponed it for some point later down the road.

There is a new project from Microsoft that does true binding to the database. Called Azure Data API Builder which under the hood uses Hot Chocolate but does full projections.

Might this be an alternative for you?

For us projections at this point has no priority for this year as we are busy tackling other big topics like AOT, Fusion, Authorization, and related topics.

We still have the larger overhaul of projections on our mind but pushed it back to later this year beginning of next year. Not sure yet. Depends also always on the need.

**Comment by michaelstaib** [comment]
You also could start a community project bringing in the missing parts... with type interceptors HC is highly extendable.

**Comment by aradalvand** [comment]
> You also could start a community project bringing in the missing parts... with type interceptors HC is highly extendable.
If you do so we can help guide you to get this working.

Yeah I'm thinking about doing that.
But some lower-level things aren't documented that well and so I need help as you said, where should I ask questions? The Slack channel?

**Comment by michaelstaib** [comment]
Slack channel is a good place ... we can have a chat together with pascal and charter a way forward.

**Comment by kresimirlesic** [comment]
Hey everyone, was any progress made on this topic?

**Comment by aradalvand** [comment]
Hi @kresimirlesic, see the latest dicussion at https://github.com/ChilliCream/graphql-platform/issues/6002 and specifically [this comment](https://github.com/ChilliCream/graphql-platform/issues/6002#issuecomment-1548258880) of mine.
I put together a plugin-esque thing that accomplishes all of this, which I'm actually satisfied with — you can check it out [here](https://github.com/aradalvand/hc-ef-custom).

**Comment by michaelstaib** [comment]
Reproduced on main as of 2026-02-25 with an EF intermediary DTO projection: querying only author.name still selected extra child columns.

Opened a fix PR with a regression test: https://github.com/ChilliCream/graphql-platform/pull/9205
